File: idea-plugin/src/main/java/com/zzf/rikki/idea/llm/LiteLlmClient.java
package com.zzf.rikki.idea.llm;

import com.zzf.rikki.idea.completion.CompletionClient;
import com.zzf.rikki.idea.completion.CompletionConfigResolver;
import com.zzf.rikki.idea.settings.RikkiSettings;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.nio.charset.StandardCharsets;

public final class LiteLlmClient implements CompletionClient {
    private static final String CHAT_SYSTEM_PROMPT = """
            You are a code completion engine. Complete the code at <|CURSOR|>.
            Rules:
            - Output ONLY the raw completion text. Nothing else.
            - No explanation. No markdown. No backticks. No prose.
            - Stop at a natural completion point (end of statement, block, or function).
            - Match the indentation and style of the surrounding code.
            """.trim();
    private static final LiteLlmClient INSTANCE = new LiteLlmClient();

    private LiteLlmClient() {
    }

    public static LiteLlmClient getInstance() {
        return INSTANCE;
    }

    public static void streamCompletion(
            String prefix,
            String suffix,
            String language,
            CompletionClient.TokenConsumer onToken
    ) throws Exception {
        RikkiSettings.State state = RikkiSettings.getInstance().getState();
        INSTANCE.streamCompletion(CompletionConfigResolver.resolve(state), prefix, suffix, language, onToken);
    }

    @Override
    public void streamCompletion(
            CompletionConfigResolver.CompletionConfig config,
            String prefix,
            String suffix,
            String language,
            TokenConsumer onToken
    ) throws Exception {
        if (!config.enabled()) {
            return;
        }
        if ((config.apiKey() == null || config.apiKey().isBlank()) && config.requiresApiKey()) {
            return;
        }
        String baseUrl = trimTrailingSlash(config.baseUrl());
        if (config.useFim()) {
            streamFim(baseUrl, config.apiKey(), config.model(), prefix, suffix, onToken);
        } else {
            streamChat(baseUrl, config.apiKey(), config.model(), prefix, suffix, language, onToken);
        }
    }

    private void streamFim(
            String baseUrl,
            String apiKey,
            String model,
            String prefix,
            String suffix,
            TokenConsumer onToken
    ) throws Exception {
        HttpURLConnection connection = openConnection(baseUrl + "/completions", apiKey);
        if (connection == null) {
            return;
        }
        try {
            connection.getOutputStream().write(buildFimBody(model, prefix, suffix).getBytes(StandardCharsets.UTF_8));
            if (connection.getResponseCode() < 200 || connection.getResponseCode() > 299) {
                return;
            }
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (!line.startsWith("data:")) {
                        continue;
                    }
                    String data = line.substring("data:".length()).trim();
                    if (data.isBlank() || "[DONE]".equals(data)) {
                        continue;
                    }
                    String token = extractFimText(data);
                    if (token != null && !token.isEmpty()) {
                        onToken.onToken(token);
                    }
                }
            }
        } finally {
            connection.disconnect();
        }
    }

    private void streamChat(
            String baseUrl,
            String apiKey,
            String model,
            String prefix,
            String suffix,
            String language,
            TokenConsumer onToken
    ) throws Exception {
        HttpURLConnection connection = openConnection(baseUrl + "/chat/completions", apiKey);
        if (connection == null) {
            return;
        }
        try {
            connection.getOutputStream().write(buildChatBody(model, prefix, suffix, language).getBytes(StandardCharsets.UTF_8));
            if (connection.getResponseCode() < 200 || connection.getResponseCode() > 299) {
                return;
            }
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (!line.startsWith("data:")) {
                        continue;
                    }
                    String data = line.substring("data:".length()).trim();
                    if (data.isBlank() || "[DONE]".equals(data)) {
                        continue;
                    }
                    String token = extractChatContent(data);
                    if (token != null && !token.isEmpty()) {
                        onToken.onToken(token);
                    }
                }
            }
        } finally {
            connection.disconnect();
        }
    }

    public static String buildFimBody(String model, String prefix, String suffix) {
        return "{\"model\":\"" + escapeJson(model) + "\",\"stream\":true,\"max_tokens\":128,\"temperature\":0.0," +
                "\"prompt\":\"" + escapeJson(prefix) + "\",\"suffix\":\"" + escapeJson(suffix) + "\"}";
    }

    public static String extractFimText(String json) {
        String key = "\"text\":\"";
        int start = json.indexOf(key);
        if (start < 0) {
            return null;
        }
        return emptyToNull(extractEscapedString(json, start + key.length()));
    }

    public static String buildChatBody(String model, String prefix, String suffix, String language) {
        String user = "Language: " + escapeJson(language) + "\\n\\n" + escapeJson(prefix) + "<|CURSOR|>" + escapeJson(suffix);
        return "{\"model\":\"" + escapeJson(model) + "\",\"stream\":true,\"max_tokens\":128,\"temperature\":0.1," +
                "\"messages\":[{\"role\":\"system\",\"content\":\"" + escapeJson(CHAT_SYSTEM_PROMPT) + "\"}," +
                "{\"role\":\"user\",\"content\":\"" + user + "\"}]}";
    }

    public static String extractChatContent(String json) {
        if (json.contains("\"content\":null")) {
            return null;
        }
        String key = "\"content\":\"";
        int start = json.indexOf(key);
        if (start < 0) {
            return null;
        }
        return emptyToNull(extractEscapedString(json, start + key.length()));
    }

    public static String extractEscapedString(String json, int start) {
        StringBuilder builder = new StringBuilder();
        int index = start;
        while (index < json.length()) {
            char c = json.charAt(index);
            if (c == '\\' && index + 1 < json.length()) {
                char escaped = json.charAt(index + 1);
                switch (escaped) {
                    case 'n' -> builder.append('\n');
                    case 'r' -> builder.append('\r');
                    case 't' -> builder.append('\t');
                    case '"' -> builder.append('"');
                    case '\\' -> builder.append('\\');
                    default -> builder.append(c);
                }
                index += 2;
                continue;
            }
            if (c == '"') {
                break;
            }
            builder.append(c);
            index++;
        }
        return builder.toString();
    }

    private HttpURLConnection openConnection(String url, String apiKey) {
        try {
            HttpURLConnection connection = (HttpURLConnection) URI.create(url).toURL().openConnection();
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/json");
            if (apiKey != null && !apiKey.isBlank()) {
                connection.setRequestProperty("Authorization", "Bearer " + apiKey);
            }
            connection.setDoOutput(true);
            connection.setConnectTimeout(5_000);
            connection.setReadTimeout(30_000);
            return connection;
        } catch (Exception ignored) {
            return null;
        }
    }

    private static String escapeJson(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private static String trimTrailingSlash(String value) {
        return value == null ? "" : value.trim().replaceAll("/+$", "");
    }

    private static String emptyToNull(String value) {
        return value == null || value.isEmpty() ? null : value;
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/llm/LiteLlmClient.kt
package com.zzf.rikki.idea.llm

import com.zzf.rikki.idea.settings.RikkiSettings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URI
import java.nio.charset.StandardCharsets

/**
 * Lightweight LLM client for inline code completion.
 *
 * Supports two modes determined by [RikkiSettings.State.completionUsesFim]:
 *  - **FIM** (Fill-In-Middle) — DeepSeek beta `/completions`, Ollama `/v1/completions`.
 *    Sends `prompt`/`suffix` fields; streaming delta field is `choices[0].text`.
 *  - **Chat** — OpenAI-compatible `/chat/completions` (OpenAI, Gemini, Moonshot, …).
 *    Sends a messages array; streaming delta field is `choices[0].delta.content`.
 */
object LiteLlmClient {

    private val CHAT_SYSTEM_PROMPT = """
        You are a code completion engine. Complete the code at <|CURSOR|>.
        Rules:
        - Output ONLY the raw completion text. Nothing else.
        - No explanation. No markdown. No backticks. No prose.
        - Stop at a natural completion point (end of statement, block, or function).
        - Match the indentation and style of the surrounding code.
    """.trimIndent()

    /**
     * Streams completion tokens into [onToken].
     * Automatically selects FIM or chat mode based on the saved completion provider.
     */
    suspend fun streamCompletion(
        prefix: String,
        suffix: String,
        language: String,
        onToken: suspend (String) -> Unit
    ) = withContext(Dispatchers.IO) {
        val s      = RikkiSettings.getInstance().state
        val apiKey = s.completionEffectiveApiKey()
        val prov   = s.completionEffectiveProvider()
        if (apiKey.isBlank() && prov != "OLLAMA") return@withContext

        val model   = s.completionEffectiveModel()
        val baseUrl = s.completionEffectiveBaseUrl().trimEnd('/')

        if (s.completionUsesFim()) {
            streamFim(baseUrl, apiKey, model, prefix, suffix, onToken)
        } else {
            streamChat(baseUrl, apiKey, model, prefix, suffix, language, onToken)
        }
    }

    // ── FIM mode (/completions, choices[0].text) ──────────────────────────────

    private suspend fun streamFim(
        baseUrl: String,
        apiKey: String,
        model: String,
        prefix: String,
        suffix: String,
        onToken: suspend (String) -> Unit
    ) {
        val conn = openConnection("$baseUrl/completions", apiKey) ?: return
        try {
            conn.outputStream.use { it.write(buildFimBody(model, prefix, suffix).toByteArray(StandardCharsets.UTF_8)) }
            if (conn.responseCode !in 200..299) return
            BufferedReader(InputStreamReader(conn.inputStream, StandardCharsets.UTF_8)).use { br ->
                var line: String?
                while (br.readLine().also { line = it } != null) {
                    currentCoroutineContext().ensureActive()
                    if (!line!!.startsWith("data:")) continue
                    val data = line!!.removePrefix("data:").trim()
                    if (data == "[DONE]" || data.isEmpty()) continue
                    val token = extractFimText(data) ?: continue
                    if (token.isNotEmpty()) onToken(token)
                }
            }
        } finally {
            conn.disconnect()
        }
    }

    private fun buildFimBody(model: String, prefix: String, suffix: String): String =
        """{"model":"${model.escapeJson()}","stream":true,"max_tokens":128,"temperature":0.0,""" +
        """"prompt":"${prefix.escapeJson()}","suffix":"${suffix.escapeJson()}"}"""

    /** Extracts text from a FIM streaming chunk: `choices[0].text`. */
    private fun extractFimText(json: String): String? {
        val key   = "\"text\":\""
        val start = json.indexOf(key).takeIf { it >= 0 }?.plus(key.length) ?: return null
        return extractEscapedString(json, start).ifEmpty { null }
    }

    // ── Chat mode (/chat/completions, choices[0].delta.content) ───────────────

    private suspend fun streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        prefix: String,
        suffix: String,
        language: String,
        onToken: suspend (String) -> Unit
    ) {
        val conn = openConnection("$baseUrl/chat/completions", apiKey) ?: return
        try {
            conn.outputStream.use { it.write(buildChatBody(model, prefix, suffix, language).toByteArray(StandardCharsets.UTF_8)) }
            if (conn.responseCode !in 200..299) return
            BufferedReader(InputStreamReader(conn.inputStream, StandardCharsets.UTF_8)).use { br ->
                var line: String?
                while (br.readLine().also { line = it } != null) {
                    currentCoroutineContext().ensureActive()
                    if (!line!!.startsWith("data:")) continue
                    val data = line!!.removePrefix("data:").trim()
                    if (data == "[DONE]" || data.isEmpty()) continue
                    val token = extractChatContent(data) ?: continue
                    if (token.isNotEmpty()) onToken(token)
                }
            }
        } finally {
            conn.disconnect()
        }
    }

    private fun buildChatBody(model: String, prefix: String, suffix: String, language: String): String {
        val user = "Language: ${language.escapeJson()}\\n\\n${prefix.escapeJson()}<|CURSOR|>${suffix.escapeJson()}"
        return """{"model":"${model.escapeJson()}","stream":true,"max_tokens":128,"temperature":0.1,""" +
               """"messages":[{"role":"system","content":"${CHAT_SYSTEM_PROMPT.escapeJson()}"},""" +
               """{"role":"user","content":"$user"}]}"""
    }

    /** Extracts content from a chat streaming delta: `choices[0].delta.content`. */
    private fun extractChatContent(json: String): String? {
        if (json.contains("\"content\":null")) return null
        val key   = "\"content\":\""
        val start = json.indexOf(key).takeIf { it >= 0 }?.plus(key.length) ?: return null
        return extractEscapedString(json, start).ifEmpty { null }
    }

    // ── Shared helpers ────────────────────────────────────────────────────────

    private fun openConnection(url: String, apiKey: String): HttpURLConnection? = try {
        (URI.create(url).toURL().openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json")
            if (apiKey.isNotBlank()) setRequestProperty("Authorization", "Bearer $apiKey")
            doOutput       = true
            connectTimeout = 5_000
            readTimeout    = 30_000
        }
    } catch (_: Exception) { null }

    /** Reads a JSON-escaped string starting at [start], stopping at the closing `"`. */
    private fun extractEscapedString(json: String, start: Int): String {
        val sb = StringBuilder()
        var i  = start
        while (i < json.length) {
            val c = json[i]
            if (c == '\\' && i + 1 < json.length) {
                when (json[i + 1]) {
                    'n'  -> { sb.append('\n'); i += 2 }
                    'r'  -> { sb.append('\r'); i += 2 }
                    't'  -> { sb.append('\t'); i += 2 }
                    '"'  -> { sb.append('"');  i += 2 }
                    '\\' -> { sb.append('\\'); i += 2 }
                    else -> { sb.append(c);    i++ }
                }
            } else if (c == '"') {
                break
            } else {
                sb.append(c); i++
            }
        }
        return sb.toString()
    }

    private fun String.escapeJson() = replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
}
